Diophantine update

smichr opened the pull request:
100% coverage with 1 pragma-no-cover.

fixes #8943
fixes #9538
closes #9539
fixes #10999 
fixes #11000 
fixes #11016 
fixes #11021 
fixes #11026
fixes #11049
fixes #11050

smichr committed:
handle partitions(0)

smichr committed:
add extra line in continued_fraction.py

smichr committed:
handle Number input to clear_coefficients

smichr committed:
diop: imports and helpers

smichr committed:
diop: diophantine

smichr committed:
diop: merge_solution

smichr committed:
diop: diop_solve

smichr committed:
diop: classify_diop

smichr committed:
diop: diop_linear

smichr committed:
diop: del extended_euclid; modify divisible

smichr committed:
diop: quadratic

smichr committed:
diop: diop_DN

smichr committed:
diop: cornacchia

smichr committed:
diop: diop_bf_DN

smichr committed:
diop: length

smichr committed:
diop: transformation_to_DN

smichr committed:
diop: find_DN

smichr committed:
diop: check_param

smichr committed:
diop: ternary

smichr committed:
diop: transformation to normal

smichr committed:
diop: parametrize_ternary_quadratic

smichr committed:
diop: ternary quadratic

smichr committed:
diop: sqf_normal

smichr committed:
diop: square_factor

smichr committed:
diop: reconstruct

smichr committed:
diop: descent and ldescent

smichr committed:
diop: gaussian_reduce

smichr committed:
diop: holzer

smichr committed:
diop: general_pythagorean

smichr committed:
diop: general_sum_of_squares

smichr committed:
diop: partition

smichr committed:
diop: sum of three|four squares

smichr committed:
diop: power representation

smichr committed:
diop: rst docs

smichr committed:
diop: tests

smichr committed:
diop: as_coefficients_dict is already defaultdict(int)

jksuom commented on sympy/utilities/iterables.py:
```diff
@@ -1381,6 +1381,12 @@ def partitions(n, m=None, k=None, size=False):
         raise ValueError("m must be > 0")
+    if n == 0:
+        if size:
+            yield 1, {0: 1}
```
I'm not sure about these. Normally partitions are expected to consist of positive integers. Maybe empty dictionary could be returned?

smichr replied:
I would feel better about not even allowing partitions of zero. What do you think about changing the error above to be "n must be > 0"?

jksuom replied:
I think that is the best solution. There is no real reason to consider partitions of zero.

smichr replied:
@jksuom, I just learned that the convention is to return count of 1 if a partition cannot be done -- one empty set -- and SymPy follows that convention:

```
>>> from sympy.functions.combinatorial.numbers import nT
>>> nT(0, 0)
1
```

So perhaps it's best to return an empty dictionary as you said so `sum(1 for 1 in partitions(n,k))` will return the right value regardless of `n` and `k`.

asmeurer commented:
Don't know what your aims are here, but I have found a [couple of issues](https://github.com/sympy/sympy/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+author%3Aasmeurer+diophantine) I've found with diophantine in the past that don't appear to be addressed here.

smichr committed:
diop: test type errors

smichr committed:
partitions(n<=0, ..) and partitions(.., m<=0) forbidden

smichr committed:
diop: 8943 (null default); affirm 9538

smichr committed:
diop: rst doc edits

smichr committed:
diop: do not use l as variable; use tuple unpacking in loops

smichr committed:
diop: handle rational expressions

smichr committed:
diop: prime_as_sum_of_two_squares

smichr committed:
diop: general_sum_of_even_powers; sum_of_squares; sum_of_powers

smichr committed:
9538: diophantine allows order of variables to be given

smichr committed:
include 0s in general sum of squares results

smichr commented:
> I have found a couple of issues

Those (and others) have now been addressed.

smichr committed:
diop: diop_quadratic (2)

smichr committed:
sympify diophantine result

smichr committed:
remove check_integrality

smichr committed:
replace broken schorn link

smichr commented:
ping @thilinarmtb or anyone else that would give this a once over. Many issues are fixed. If there is no objection, I will commit in 96 hours.

thilinarmtb commented:
@smichr : I'll go through this during the weekend.

smichr commented:
@thilinarmtb , thanks. One thing I changed was the use of Integer vs int. Only Diophantine returns Integers; all other routines return an int. I know there was some talk about only exposing Diophantine. That seems like a good idea to me. I think I would also expose `power_representation` and/or the alias `sum_of_Powers`) and `sum_of_squares`.

I am also not sure about the need of keeping `sum_of_three_squares` and `sum_of_four_squares` except that if someone just wants a single instance, these are a quick way of getting it.

Finally, I tested loops over integers seeing when it might be more advantageous to use "smart" writing as squares instead of the backtracking recursive approach that is used by `power_representation`. I corrected the routine to watch for cases like writing 9 squares as a square and *two groups of four*, even wrote my own routines that did essentially the same thing, but it always seemed faster to just use the recursive approach. I'm a little doubtful that it truly is the best way to go; it seems like at some point the recursive approach would get slow and the "schorn" reference suggests the same, but everything I tested seemed to show that the recursive approach used here is best.

smichr committed:
rational=True option for as_coeff_Add; use in clear_coefficients

thilinarmtb commented on sympy/solvers/diophantine.py:
```diff
@@ -1827,65 +1839,52 @@ def _transformation_to_normal(var, coeff):
-                    T.row_swap(0, 2)
-                    T.col_swap(0, 2)
-                    return T
+    elif coeff[y*z] != 0:
```
I guess we can simplify this as in other cases as well.

smichr replied:
I'm sorry, I'm not sure in which way you mean this might be simplified.

thilinarmtb commented on sympy/solvers/diophantine.py:
```diff
@@ -2062,58 +2062,61 @@ def _diop_ternary_quadratic_normal(var, coeff):
-    free.
+    Return `a', b', c'`, the coefficients of the square-free normal
+    form of `ax^2 + by^2 + cz^2 = 0`, where `a', b', c'` are pairwise
+    prime and `\gcd(a', b', c')` is `1`.  If `steps` is True then
```
if `a, b, c` are pairwise prime, then `gcd(a, b, c) ` is 1. Do we need to restate it?

smichr replied:
I removed that phrase now.

thilinarmtb commented on sympy/solvers/diophantine.py:
```diff
@@ -2262,8 +2261,6 @@ def descent(A, B):
     if A == 1:
         return (1, 1, 0)
-    if B == -1:
-        return (None, None, None)
```
Why is this line removed?

smichr replied:
It is already handled by the routine. But the case of `A = B = -1` leads to infinite recursion without that; I added it back with a test.

thilinarmtb commented on sympy/solvers/diophantine.py:
```diff
@@ -2341,33 +2341,67 @@ def norm(u, w, a, b):
+        y = (y_0*A - 2*v*B) / k
+        z = (z_0*A - 2*w*B) / k
+        assert all(i.is_Integer for i in (x, y, z))
+        step += 1
```
Looks like this is a different implementation  than the existing one.

smichr replied:
Yes, there were some issues with the former implementation. I just reinterpreted the reference and what I've committed is that interpretation.  Here are some examples of failures in the former version:

```
>>> f = lambda eq: _diop_ternary_quadratic_normal((x, y, z), eq.as_coefficients_dict())
>>> f(92*x**2 - 99*y**2 - z**2)
(28557, 21296, 155058)  <-- should be set([(9, 7, 51)])
>>> f(234*x**2 - 65601*y**2 - z**2)
(5945, 31, 126819) <-- should be set([(233, 1, 3555)])
```

thilinarmtb replied:
Great !!!

thilinarmtb commented on doc/src/modules/solvers/diophantine.rst:
```diff
@@ -281,7 +281,7 @@ level API.
 
-If you want to get a more thorough idea about the the Diophantine module please
+If you want to get a more thorough understanding of the Diophantine module please
 refer to the following blog.
```
I guess `refer the following blog` sounds better.

smichr replied:
It sounds fine as it is (to me). Do you want me to shorten it further to "More discussion about the Diophantine module can be found at the following blog:"?

thilinarmtb commented:
@smichr : I went through the PR and you have done an amazing job of polishing the code and documentation. Thanks for your work ! I left a few comments.

> I know there was some talk about only exposing Diophantine. That seems like a good idea to me. I think I would also expose power_representation and/or the alias sum_of_powers) and sum_of_squares.

Yes, if we expose other routines, there can be a lot of confusion. Why would you expose the sum of powers?

> I am also not sure about the need of keeping sum_of_three_squares and sum_of_four_squares except that if someone just wants a single instance, these are a quick way of getting it.

IMO, we shouldn't expose these two. `diop_general_sum_of_squares` check for these two cases upfront so calling them directly won't have any significant impact.

> Finally, I tested loops over integers seeing when it might be more advantageous to use "smart" writing as squares instead of the backtracking recursive approach that is used by power_representation. I corrected the routine to watch for cases like writing 9 squares as a square and two groups of four, even wrote my own routines that did essentially the same thing, but it always seemed faster to just use the recursive approach. I'm a little doubtful that it truly is the best way to go; it seems like at some point the recursive approach would get slow and the "schorn" reference suggests the same, but everything I tested seemed to show that the recursive approach used here is best.

I see. When I implemented it, even I expected it to be slow for larger values.

smichr commented:
I found a reference for counting the ways to write `n` as a sum of 2, 3 or 4 squares [here](http://web.maths.unsw.edu.au/~mikeh/webpapers/paper73.pdf) but I don't know what the sigma function is in equation (6). Can you help me? I also found a reference that refers to the [Jacobi formula](http://kam.mff.cuni.cz/~klazar/ostrava1.pdf). Is sigma(n) the sum of the divisors of n?

thilinarmtb commented:
Yes, I think it is the sum of divisors of n.

Take the last formula in the page 3 of your first [reference](http://web.maths.unsw.edu.au/~mikeh/webpapers/paper73.pdf) and the equation 1 in the second [reference] (http://kam.mff.cuni.cz/~klazar/ostrava1.pdf) (Jacobi formula paper). They are the same as LHSs are the same in both the equations (Only the notation is different, `\phi(q)` instead of the summation).

Lemma 4 [here](http://www.rowan.edu/colleges/csm/departments/math/facultystaff/osler/89%20Surprising%20Connection%20Between%20Partitions%20and%20Divisors.pdf.pdf), indicates that `\sigma(n)` in the first reference is indeed the sum of divisors. Equation 1 in the second reference is just a rewrite of the equation in the first reference using the Lemma 4.

smichr committed:
diop: test for A==B==-1 in ldescent

The test for A == -1 is implied by the logic which precedes
the added lines. We know that |A| <= |B| and we know that A
is not 1 at this, so if B == -1, A must be -1.

smichr committed:
diop: make sure parametrized soln is returned

smichr committed:
note which test addressed 11049

smichr committed:
expose only classify_diop and diophantine

smichr committed:
classify_diop defaults to dict for coeff

smichr committed:
rename diop test

smichr committed:
describe use of syms in diophantine docstring

smichr committed:
skip factoring for cases where it is not needed

smichr committed:
remove use of Integer; add partition tests

smichr commented:
> diop_general_sum_of_squares check for these two cases upfront

This no longer uses those routines since they only returned one particular instance; now all solutions are returned.

Regarding `partition`: you found a great reference for a routine but that aspect of the original `partition` function was not used (the case when all partitions should be generated) and the issue I opened indicated that there were problems. As a quick fix, I just replaced it with something that would expand the dictionaries received from `partitions`; I have since found a nice way to use that reference, however, and will update this PR with it. Alternatively, `sum_of_powers` could be made to return a dictionary, too, and then it could just use the `partitions` function directly.

thilinarmtb commented:
That's great !! Which reference are you talking about? Did I implement that wrong?

smichr commented:
> Which reference are you talking about? Did I implement that wrong?

The reference is [this one](http://jeromekelleher.net/generating-integer-partitions.html); there must be some problem (see #11050).

smichr commented:
> I expected it to be slow for larger values

I think part of the reason it is faster than expected is because you are stepping like  int(sqrt(10)), int(sqrt(10//2)), int(sqrt(10//3)), ... rather than blindly starting from int(sqrt(10)) and going down by 1 each time. As that n=10 gets to be bigger, this has a greater impact.

smichr commented:
I'm just recording here the routines for counting the ways to write n as a sum of 2, 3 or 4 squares (including zeros). The docstrings need the correct imports if this is added to SymPy:

```python
http://web.maths.unsw.edu.au/~mikeh/webpapers/paper73.pdf
from sympy.ntheory.primetest import is_square
def delta(n, den=None):
    if den:
        n, r = divmod(n, den)
        if r:
            return 0
    return is_square(n)

def d(r, m, n):
    return sum(1 for di in divisors(n) if di % m == r)

def e(r, s, m, n, den=None):
    if den and n % den:
        return 0
    return d(r, m, n) - d(s, m, n)

def p2(n):
    """Count the ways to write n as a sum of 2 or fewer squares.

    Examples
    ========

    >>> [p2(i) for i in range(1,18)]
    [1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1]
    >>> [len(list(sos(i,2,1))) for i in range(1,18)]
    [1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1]
    """
    return (e(1, 3, 4, n) + delta(n) + delta(n, 2))//2

def p3(n):
    """Count the ways to write n as 3 or fewer squares.

    Examples
    ========

    >>> [p3(i) for i in range(1,18)]
    [1, 1, 1, 1, 1, 1, 0, 1, 2, 1, 1, 1, 1, 1, 0, 1, 2]
    >>> [len(list(sos(i,3,1))) for i in range(1,18)]
    [1, 1, 1, 1, 1, 1, 0, 1, 2, 1, 1, 1, 1, 1, 0, 1, 2]
    """
    return (
        4*e(1, 3, 4, n) + 
        2*sum(e(1,3,4,n-k**2) for k in range(1, integer_nthroot(n, 2)[0] + 1)) + 
        3*e(1, 7, 8, n) + 
        3*e(3, 5, 8, n) + 
        5*delta(n) + 
        3*delta(n, 2) + 
        4*delta(n, 3))//12

def sigma(n):
    return sum(divisors(n, generator=True))

def sigma4(n):
    return sum(d for d in divisors(n, generator=True) if d % 4)

def sigma_not4(n):
    '''for i in range(1,10000):
        assert sigma_not4(i)==sigma4(i)
        assert sigma_not4(i)==sigma(i) - (4*sigma(i/4) if not i%4 else 0)
    '''
    if not n % 4:
        m = multiplicity(4, n)
        if m:
            n //= 4**m
            if n % 2:
                n *= 2
    return sum(divisors(n, generator=True))

def p4(n):
    """Count the ways to write n as 4 or less squares.

    Examples
    ========

    >>> [p4(i) for i in range(1,18)]
    [1, 1, 1, 2, 1, 1, 1, 1, 2, 2, 1, 2, 2, 1, 1, 2, 2]
    >>> [len(list(sos(i, 4, zeros=True))) for i in range(1,18)]
    [1, 1, 1, 2, 1, 1, 1, 1, 2, 2, 1, 2, 2, 1, 1, 2, 2]
    """
    N = integer_nthroot(n, 2)[0]
    N2 = integer_nthroot(n//2, 2)[0]
    return (
        sigma_not4(n) +
        17*e(1, 3, 4, n) + 
        4*sum(e(1,3,4,n - k**2) for k in range(1, N + 1)) +
        12*sum(e(1,3,4,n - 2*k**2) for k in range(1, N2 + 1)) +
        6*e(1, 7, 8, n) + 
        6*e(3, 5, 8, n) + 
        6*e(1, 3, 4, n, 2) + 
        8*e(1, 2, 3, n) + 
        16*e(1, 2, 3, n, 4) + 
        16*delta(n) + 
        12*delta(n, 2) + 
        8*delta(n, 3) + 
        12*delta(n, 4))//48
```

smichr committed:
iterables: add ordered_partitions; use in diophantine

smichr committed:
diop: encode permanent coverage for sum_of_four_squares

smichr commented:
ping @thilinarmtb  -- ok, the new partitions is in place.

thilinarmtb commented:
Great ! +1 to merge as it is.

smichr committed:
partitions returns empty dict for emptyset result

smichr committed:
ordered_partitions returns [] for emptyset result

smichr committed:
diop: fix imports

smichr commented:
@thilinarmtb , did you write the `power_representation` routine? What's the backstory -- if you wrote it, was it just the most obvious way to you to solve the problem?
